File: android/app/src/main/java/com/sapiens/gagaodok/sync/SyncRoomExposurePolicy.kt
package com.sapiens.gagaodok.sync

object SyncRoomExposurePolicy {
    fun isVisible(originSpaceId: String, viewerSpaceId: String): Boolean = when {
        originSpaceId == "MAC_SPACE" && viewerSpaceId == "PHONE_SPACE" -> true
        originSpaceId == "TABLET_SPACE" && viewerSpaceId == "MAC_SPACE" -> true
        originSpaceId == "TABLET_SPACE" && viewerSpaceId == "PHONE_SPACE" -> true
        else -> false
    }
}

/** Read-only UI boundary with no ChatStore or local conversation path. */
class SyncRemoteRoomCatalog(
    private val repository: SyncRemoteRoomRepository,
    private val viewerSpaceId: String,
) {
    fun refresh(): List<SyncRemoteRoomSnapshot> = repository.list()
        .filter { SyncRoomExposurePolicy.isVisible(it.handle.originSpaceId, viewerSpaceId) }
        .sortedWith(
            compareByDescending<SyncRemoteRoomSnapshot> { it.messages.lastOrNull()?.timestamp ?: "" }
                .thenBy { it.handle.roomId },
        )

    fun open(handle: SyncRoomHandle): SyncRemoteRoomSnapshot? {
        if (!SyncRoomExposurePolicy.isVisible(handle.originSpaceId, viewerSpaceId)) return null
        return repository.load(handle)
    }
}

File: Sources/KakaoSapiens/Services/SyncRoomExposurePolicy.swift
import Foundation

public enum SyncRoomExposurePolicy {
    /// Product decisions 3 and 5. Own-space rooms remain in the existing local
    /// list, so they are deliberately false here.
    public static func isVisible(originSpaceID: String, viewerSpaceID: String) -> Bool {
        switch (originSpaceID, viewerSpaceID) {
        case ("MAC_SPACE", "PHONE_SPACE"),
             ("TABLET_SPACE", "MAC_SPACE"),
             ("TABLET_SPACE", "PHONE_SPACE"):
            return true
        default:
            return false
        }
    }
}

/// A read-only boundary used by the UI. It can list and open the dedicated
/// remote repository, but has no reference to ChatRoomManager or local files.
public struct SyncRemoteRoomCatalog {
    private let repository: SyncRemoteRoomRepository
    private let viewerSpaceID: String

    public init(repository: SyncRemoteRoomRepository, viewerSpaceID: String) {
        self.repository = repository
        self.viewerSpaceID = viewerSpaceID
    }

    public func refresh() throws -> [SyncRemoteRoomSnapshot] {
        try repository.list()
            .filter {
                SyncRoomExposurePolicy.isVisible(
                    originSpaceID: $0.handle.originSpaceID,
                    viewerSpaceID: viewerSpaceID
                )
            }
            .sorted {
                let left = $0.messages.last?.timestamp ?? ""
                let right = $1.messages.last?.timestamp ?? ""
                if left != right { return left > right }
                return $0.handle.roomID.uuidString < $1.handle.roomID.uuidString
            }
    }

    public func open(_ handle: SyncRoomHandle) throws -> SyncRemoteRoomSnapshot? {
        guard SyncRoomExposurePolicy.isVisible(
            originSpaceID: handle.originSpaceID,
            viewerSpaceID: viewerSpaceID
        ) else { return nil }
        return try repository.load(handle)
    }
}
